pkg/controller/nodeipam/ipam/multicidrset/multi_cidr_set.go

```diff
@@ -0,0 +1,365 @@
+		}
+	default:
+		return nil, fmt.Errorf("invalid IP: %s", s.ClusterCIDR.IP)
+	}
+	return &net.IPNet{
+		IP:   ip,
+		Mask: s.nodeMask,
+	}, nil
+}
+
+// NextCandidate returns the next candidate and the last evaluated index
+// for the current cidrSet. Returns nil if the candidate is already allocated.
+func (s *MultiCIDRSet) NextCandidate() (*net.IPNet, int, error) {
+	s.Lock()
+	defer s.Unlock()
+
+	if s.allocatedCIDRs == s.MaxCIDRs {
+		return nil, 0, &CIDRRangeNoCIDRsRemainingErr{
+			CIDR: s.Label,
+		}
+	}
+
+	candidate := s.nextCandidate
+	for i := 0; i < s.MaxCIDRs; i++ {
+		nextCandidateCIDR, err := s.indexToCIDRBlock(candidate)
+		if err != nil {
+			return nil, i, err
+		}
+		// Check if the nextCandidate is not already allocated.
+		if _, ok := s.AllocatedCIDRMap[nextCandidateCIDR.String()]; !ok {
+			s.nextCandidate = (candidate + 1) % s.MaxCIDRs
+			return nextCandidateCIDR, i, nil
+		}
+		candidate = (candidate + 1) % s.MaxCIDRs
+	}
+
+	return nil, s.MaxCIDRs, &CIDRRangeNoCIDRsRemainingErr{
+		CIDR: s.Label,
+	}
+}
+
+func (s *MultiCIDRSet) getBeginningAndEndIndices(cidr *net.IPNet) (int, int, error) {
+	if cidr == nil {
+		return -1, -1, fmt.Errorf("error getting indices for cluster cidr %v, cidr is nil", s.ClusterCIDR)
+	}
+	begin, end := 0, s.MaxCIDRs-1
+	cidrMask := cidr.Mask
+	maskSize, _ := cidrMask.Size()
+	var ipSize int
+
+	if !s.ClusterCIDR.Contains(cidr.IP.Mask(s.ClusterCIDR.Mask)) && !cidr.Contains(s.ClusterCIDR.IP.Mask(cidr.Mask)) {
+		return -1, -1, fmt.Errorf("cidr %v is out the range of cluster cidr %v", cidr, s.ClusterCIDR)
+	}
+
+	if s.clusterMaskSize < maskSize {
+		var err error
+		ipSize = net.IPv4len
+		if cidr.IP.To4() == nil {
+			ipSize = net.IPv6len
+		}
+		begin, err = s.getIndexForCIDR(&net.IPNet{
+			IP:   cidr.IP.Mask(s.nodeMask),
+			Mask: s.nodeMask,
+		})
+		if err != nil {
+			return -1, -1, err
+		}
+		ip := make([]byte, ipSize)
+		if cidr.IP.To4() != nil {
+			ipInt := binary.BigEndian.Uint32(cidr.IP) | (^binary.BigEndian.Uint32(cidr.Mask))
+			binary.BigEndian.PutUint32(ip, ipInt)
+		} else {
+			// ipIntLeft          |         ipIntRight
+			// 2001:0DB8:1234:0000:0000:0000:0000:0000
+			ipIntLeft := binary.BigEndian.Uint64(cidr.IP[:net.IPv6len/2]) | (^binary.BigEndian.Uint64(cidr.Mask[:net.IPv6len/2]))
+			ipIntRight := binary.BigEndian.Uint64(cidr.IP[net.IPv6len/2:]) | (^binary.BigEndian.Uint64(cidr.Mask[net.IPv6len/2:]))
+			binary.BigEndian.PutUint64(ip[:net.IPv6len/2], ipIntLeft)
+			binary.BigEndian.PutUint64(ip[net.IPv6len/2:], ipIntRight)
+		}
+		end, err = s.getIndexForCIDR(&net.IPNet{
+			IP:   net.IP(ip).Mask(s.nodeMask),
+			Mask: s.nodeMask,
+		})
+		if err != nil {
+			return -1, -1, err
+		}
+	}
+	return begin, end, nil
+}
+
+// Release releases the given CIDR range.
+func (s *MultiCIDRSet) Release(cidr *net.IPNet) error {
+	begin, end, err := s.getBeginningAndEndIndices(cidr)
+	if err != nil {
+		return err
+	}
+	s.Lock()
```

Comment:
Why put the lock down here? does s.getBeginning... take the lock?
This introduces a brief window where it isn't locked...

Reply:
No it doesn't there is no shared resource being accessed in s.getBeginning... , it is taking in a cidr and just returning the indices, similar to the current implementation: https://github.com/kubernetes/kubernetes/blob/cdc60112a676010dbc0ba5650732386e2dedebd0/pkg/controller/nodeipam/ipam/cidrset/cidr_set.go#L250-L256